core/src/server/rpc_module/methods.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::collections::hash_map::Entry;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use crate::error::Error;
+use crate::id_providers::RandomIntegerIdProvider;
+use crate::server::helpers::{BoundedSubscriptions, MethodSink};
+use crate::traits::ToRpcParams;
+use jsonrpsee_types::error::{CallError, ErrorCode, ErrorObject};
+use jsonrpsee_types::{ErrorResponse, Id, Params, Request, Response, SubscriptionId as RpcSubscriptionId};
+use rustc_hash::FxHashMap;
+use serde::de::DeserializeOwned;
+use tokio::sync::mpsc;
+
+use crate::server::helpers::MethodResponse;
+
+use super::types::*;
+
+/// Reference-counted, clone-on-write collection of synchronous and asynchronous methods.
+#[derive(Default, Debug, Clone)]
+pub struct Methods {
+	pub(crate) callbacks: Arc<FxHashMap<&'static str, MethodCallback>>,
+}
+
+impl Methods {
+	/// Creates a new empty [`Methods`].
+	pub fn new() -> Self {
+		Self::default()
+	}
+
+	/// Verify a method name is available, or return an error if it's already taken.
+	pub fn verify_method_name(&mut self, name: &'static str) -> Result<(), Error> {
+		if self.callbacks.contains_key(name) {
+			return Err(Error::MethodAlreadyRegistered(name.into()));
+		}
+
+		Ok(())
+	}
+
+	/// Inserts the method callback for a given name, or returns an error if the name was already taken.
+	/// On success it returns a mut reference to the [`MethodCallback`] just inserted.
+	pub fn verify_and_insert(
```

Comment:
there are no code change at all. this is the only public visibility change. other pub(crate) are only to make it compile.